### ISSUE TITLE
proposal: scoped identifiers

### ISSUE THREAD

**Issue by tetron** [common-workflow-language/common-workflow-language#105]
One user-unfriendly aspect of CWL is that identifiers are document
scoped, and require the "#" sigil everywhere.  This puts a burden on
CWL authors to avoid name conflicts even between unrelated parts of a
workflow, and it is inconsistent with the related usage in HTML/XML
where a section is identifier as `<p id="foo">` and the "#" sigil is
only used when making a reference:
`<a href="#foo">`.

There is a reason for this; CWL is based on JSON-LD and uses RFC 3986
URI relative reference resolution rules (which treat document
fragments as opaque) and does not have a concept of implicit scope
(you can have explicit scope by declaring nested JSON-LD contexts.)

However, CWL is not actually JSON-LD; CWL documents are YAML (not
plain JSON) and support "import" and "include" preprocessing
directives.  We can take advantage of this preprocessing step to
support scoped identifiers and make CWL documents just a little bit
easier to read and write.

Here are the proposed rules:
 - An object with an "id" sets the base URL used for resolution of
nested objects
 - A relative id that does not not start with "#" is joined with the
base URL fragment
   - If there is no fragment in the base URL, set the fragment portion
of the base URL
   - If there is a fragment in the base URL, join the base URL
fragment with the relative id, separated "/"
 - A relative id that does start with "#" remains document scoped and
replaces the fragment of the base URL.
 - Cross references remain document scoped and must start with "#"

This turns out to be a fairly small change to the existing reference
resolution code, I have pushed this to the "scoped-ids" branch on github.

To make this concrete, here is an example workflow with scoped
identifiers.  Note far fewer instances of "#" and much more uniform
naming of nested identifiers.

```yaml
id: main
class: Workflow
inputs:
  - id: infile
    type: File
  - id: term
    type: string
outputs:
  - id: outfile
    type: File
    source: "#main/search/result"

steps:
  - id: index
    run: "#index"
    inputs:
      - { id: file, source: "#main/infile" }
    outputs:
      - id: result

  - id: search
    run: "#search"
    inputs:
      - { id: file, source: "#main/index/result" }
      - { id: term, source: "#main/term" }
    outputs:
      - id: result
```

The above workflow, after preprocessing to resolve the scoped identifiers into
their actual identifiers.

```yaml
id: "#main"
class: Workflow
inputs:
  - id: "#main/infile"
    type: File
  - id: "#main/term"
    type: string
outputs:
  - id: "#main/outfile"
    type: File
    source: "#main/search/result"

steps:
  - id: "#main/index"
    run: "#index"
    inputs:
      - { id: "#main/index/file", source: "#main/infile" }
    outputs:
      - id: "#main/index/result"

  - id: "#main/search"
    run: "#search"
    inputs:
      - { id: "#main/search/file", source: "#main/index/result" }
      - { id: "#main/search/term", source: "#main/term" }
    outputs:
      - id: "#main/search/result"
```

For reference, here is the draft-2 workflow the above workflow is based on.

```yaml
id: "#main"
class: Workflow
inputs:
  - id: "#infile"
    type: File
  - id: "#term"
    type: string
outputs:
  - id: "#outfile"
    type: File
    source: "#search.search_result"

steps:
  - run: {import: "#index"}
    inputs:
      - { id: "#index.index_file", source: "#infile" }
    outputs:
      - id: "#index.index_result"

  - run: {import: "#search"}
    inputs:
      - { id: "#search.search_file", source: "#index.index_result" }
      - { id: "#search.search_term", source: "#term" }
    outputs:
      - id: "#search.search_result"
```

**Comment by dionjwa** [comment]
+1

I do not understand the reason to scope ids over the whole document. Inputs for a given step only need to be scoped to that step itself, and document global uniqueness imposes a significant burden and unnecessary complexity when implementing.

**Comment by tetron** [comment]
For reference, here's the reasoning why draft-2 specifies document level scope:

CWL is based on JSON-LD, and JSON-LD does not have an mechanism for implicit lexical scope.  It has an implicit "base URL" which is used to resolve relative identifiers, but this is generally over the whole document.  So using just the JSON-LD relative reference rules,`#foo` and `#bar` are relative document fragment identifiers which are both resolved against the same base url `stuff.cwl` to yield `stuff.cwl#foo` and `stuff.cwl#bar`.

This proposal is to change the relative identifier rules to include implicit lexical scoping, with the benefit of being much easier for humans to work with, but the drawback of being incompatible with JSON-LD without an additional preprocessing pass to rewrite the identifiers from lexically scoped to document scoped.

So the underlying reason is to work within the limitations of other standards that CWL is based on, and carefully weigh the benefits of adding new behavior against the drawbacks of that behavior being non-standard.

**Comment by ntijanic** [comment]
Since we're already doing pre-processing, this looks like it would make it easier to write the YAML files by hand (the "#" characters were annoying to write in YAML since they require quotes).

To make sure I understand, can the rules be phrased as:

 1. `id` properties that do not start with `#` and do not contain a `:` should be prefixed with `<scope>/` or `<scope>#` where `<scope>` is the "expanded" ID of object's container (the object that contains this object). `<scope>/` is used for scopes which are IDs with document fragments, and `<scope>#` is used for scopes which are not.
 2. if the containing object does not have an ID, `<scope>` should be the ID of the container's container.
 3. Repeat 2) until you reach an "expanded" ID.

Are we limiting the IDs to be only URLs (not URNs)? Seems this makes sense, since we're using fragments. People can use other properties to annotate with URNs.

**Comment by tetron** [comment]
Merged in 6e0854b5df775e8a54321ec3e803b118f9326ae1, schema salad documentation will have comprehensive rules for URI normalization.
